keps/sig-apps/4368-support-managed-by-for-batch-jobs/README.md

```diff
@@ -362,6 +365,67 @@ We may come up with more validation rules during the implementation phase.
+as well as merge [Improve the Job API comment for ready field](https://github.com/kubernetes/kubernetes/pull/125189).
+
+#### Terminating pods and terminal Job conditions
+
+During the development process of Alpha in 1.30 we considered adding a
+validation rule enforcing that the Job terminal conditions (`Failed` or `Complete`)
+are only added when all pods are terminated (`status.terminating=0` and
+`status.ready=0`). However, the rule turned out to be violated by the built-in
+Job controller (see issue [#123775](https://github.com/kubernetes/kubernetes/issues/123775)).
+
+We are going to solve this issue by delaying the addition of the Job terminal
+conditions (Failed or Complete) until the pods are terminated
+(`terminating=0` and `ready=0`).
+
+One complication of this approach is that pod termination may take an
+arbitrarily long if the non-standard pod graceful termination period
+(`terminationGracePeriodSeconds`) is configured (30s by default). In order to
+give the API clients flexibility to know the fate of a Job as soon as possible
+we extend the scope for the following Job conditions:
+- `FailureTarget` - introduced in [Pod Failure Policy KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures)
+- `SuccessCriteriaMet` - introduced in [Success Policy KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3998-job-success-completion-policy)
+
+Note that, with the new conditions an API client can know the fate of a Job
+earlier than currently, since adding the conditions will not wait for emptying
+of the `uncountedTerminatedPods` structure, as shown in the
+[experiment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2114710754).
+
+Since the fix is needed for this KEP and the
+[Pod Replacement Policy KEP](https://github.com/kubernetes/enhancements/issues/3939),
+we intend to protect the fix with the OR of the `JobManagedBy` and the
+`JobReplacementPolicy` feature gates.
+
+Additionally, fixing the issue [#123775](https://github.com/kubernetes/kubernetes/issues/123775) will
+also require fixing:
+- [Job controller reports the count of terminating pods with unnecessary delay](https://github.com/kubernetes/kubernetes/issues/125089) and
+- [Job controller reports the count of ready pods with unnecessary delay](https://github.com/kubernetes/kubernetes/issues/125185).
+
+Note also that the fix impacts `CronJob` when using the `Forbid` concurrency
+policy (see [here](https://kubernetes.io/docs/concepts/workloads/controllers/cron-jobs/#concurrency-policy)).
+However, it can be argued that the current behavior of the CronJob / Job integration
+does not match the expectations of the `Forbid` policy
+(see [comment](https://github.com/kubernetes/kubernetes/issues/123775#issuecomment-2115200217)),
+and can be handled as a bug.
+
+The validation rule will be added in the +1 release (1.32) to ensure the
```

Comment:
Since we're remaining in Alpha, it seems acceptable to me that we add the validation rule now and accept the hiccup during upgrades with the alpha enabled.  Doing this will allow those developing new job controllers to see the correct behavior in 1.31 instead of waiting until 1.32.  The cost of doing this is a few fail update/status during rolling updates of alpha enabled clusters.

Reply:
+1

Reply:
I see, it makes perfect sense. Initially I added validation to the graduation criteria for Beta, cause while nice-to-have it does not seem required for alpha.

After the comment I moved it to the second alpha section to better reflect the intention of working on it during the second alpha.